src/yunohost/firewall.py

```diff
@@ -327,66 +331,57 @@ def firewall_upnp(action="status", no_refresh=False):
 
     """
     firewall = firewall_list(raw=True)
-    enabled = firewall["uPnP"]["enabled"]
-
-    # Compatibility with previous version
-    if action == "reload":
-        logger.debug("'reload' action is deprecated and will be removed")
-        try:
-            # Remove old cron job
-            os.remove("/etc/cron.d/yunohost-firewall")
-        except Exception:
-            pass
-        action = "status"
-        no_refresh = False
 
-    if action == "status" and no_refresh:
-        # Only return current state
-        return {"enabled": enabled}
-    elif action == "enable" or (enabled and action == "status"):
+    if action == "status":
+        enabled = firewall["uPnP"]["enabled"]
+    elif action == "enable":
         # Add cron job
         with open(UPNP_CRON_JOB, "w+") as f:
             f.write(
-                "*/50 * * * * root "
+                "*/10 * * * * root "
                 "/usr/bin/yunohost firewall upnp status >>/dev/null\n"
             )
-        # Open port 1900 to receive discovery message
-        if 1900 not in firewall["ipv4"]["UDP"]:
-            firewall_allow("UDP", 1900, no_upnp=True, no_reload=True)
-            if not enabled:
-                firewall_reload(skip_upnp=True)
         enabled = True
-    elif action == "disable" or (not enabled and action == "status"):
+    elif action == "disable":
         try:
             # Remove cron job
             os.remove(UPNP_CRON_JOB)
         except Exception:
             pass
         enabled = False
-        if action == "status":
-            no_refresh = True
     else:
         raise YunohostValidationError("action_invalid", action=action)
 
-    # Refresh port mapping using UPnP
+    # Refresh port mapping
+    refresh_success = True
     if not no_refresh:
-        upnpc = miniupnpc.UPnP(localport=1)
+        # Open port to receive discovery message
+        process.run_commands(
+            ["iptables -w -A INPUT -p udp --dport %d -j ACCEPT" % SSDP_CLIENT_PORT],
+            callback=_on_rule_command_error,
+        )
+        upnpc = miniupnpc.UPnP(localport=SSDP_CLIENT_PORT)
         upnpc.discoverdelay = 3000
-
         # Discover UPnP device(s)
         logger.debug("discovering UPnP devices...")
         nb_dev = upnpc.discover()
         logger.debug("found %d UPnP device(s)", int(nb_dev))
+        # Close discovery port
+        process.run_commands(
+            ["iptables -w -D INPUT -p udp --dport %d -j ACCEPT" % SSDP_CLIENT_PORT],
```

Comment:
The current code deletes the firewall rule which opens the port, while proposed change adds a new rule which overrides the existing rule - this would cause a conflict the next time the port is opened.